### DISCUSSION TITLE
Slightly simpler "cd on quit"; running in a subshell

### DISCUSSION THREAD

**Discussion by danielittlewood0** [jarun/nnn#1733] [General]
Hi all, I've been using nnn for a long time. I'm also a long-time user of the "configure cd on quit" behaviour. IIRC this used to be done with a bash alias `n()` (which I still use from time to time) and requires messing around with a file. There are new instructions there now.

I think this can be achieved using a simple subshell, and I wonder whether anyone else uses this workflow, or if it would be helpful to mention it in the wiki. The basic idea is to use the `-p` argument, e.g.

```
cd $(nnn -p -)
vim $(nnn -p -)
```

this is similar to `vim $(fzf)`, which I've definitely seen other people use. The behaviour is slightly different because you have to select the directory to be changed to rather than merely inhabiting it. It's possible to add some error checking (in case you select nothing, or multiple things, or select a file which is not a directory).

```
dir=$(nnn -p - | head -n1) && [ -n "$dir" ] && [ -d "$dir" ] && cd "$dir"
```

This can be unworkable if you do not have execute permission of the parent of the directory you want to access, but in normal usage I have never had this happen (if I want to cd to my home directory for example, I would not even open nnn). This could be easily fixed by optionally including  `.` when nnn lists a directory. Enabling hidden files doesn't do this, on my end.

There are actually a lot more use cases when using `nnn -p -` in a subshell like this. For instance, I think "open file" only works with the currently hovered file, but it is easy to generalise it to multiple files in this way. I think that's how I came across it in the first place.

I wonder if I'm just saying something "everybody knows" or if this actually helps people using the tool.

**Comment by N-R-K** [comment]
>The basic idea is to use the -p argument

The `-p` flag IIRC has a lot of other side effects which can mess up normal selection operations. It was mainly designed for using nnn as a picker for vim and isn't suitable for regular usage.

If it works for you, then that's great, but I don't think we can recommend as basic guideline since it might cause issues.

**Reply by danielittlewood0** [reply]
@N-R-K Ok, thanks for replying! I wonder what sort of side effects you mean? Some are obvious (e.g. Enter does not open, but returns the filename instead). So far what I describe "seems" to work, but I wonder if you mean something important I should be aware of.

**Reply by N-R-K** [reply]
If you grep the `src/nnn.c` file for `g_state.picker` you will notice there are many places where nnn behaves differently when picker mode is enabled.
